lib/streamlit/runtime/fragment.py

```diff
@@ -105,9 +108,73 @@ def fragment(
 
 
 @gather_metrics("experimental_fragment")
-def fragment(  # type: ignore[empty-body]
+def fragment(
     func: F | None = None,
     *,
     run_every: float | None = None,
 ) -> Callable[[F], F] | F:
-    pass
+    """TODO(vdonato): Write a docstring for this function."""
+
+    if func is None:
+        # Support passing the params via function decorator
+        def wrapper(f: F) -> F:
+            return fragment(
+                func=f,
+                run_every=run_every,
+            )
+
+        return wrapper
+    else:
+        non_optional_func = func
+
+    @wraps(non_optional_func)
+    def wrap(*args, **kwargs):
+        import streamlit as st
+
+        ctx = get_script_run_ctx()
+        if ctx is None:
+            return
+
+        if len(dg_stack.get()) > 0:
+            dg_stack_snapshot = deepcopy(dg_stack.get())
+        else:
+            with st.container():
+                dg_stack_snapshot = deepcopy(dg_stack.get())
+
+        # TODO(vdonato): Maybe improve how we construct fragment_id hashes.
+        active_dg = dg_stack_snapshot[-1]
+        h = hashlib.new("md5")
+        h.update(
+            f"{non_optional_func.__module__}.{non_optional_func.__qualname__}{active_dg._get_delta_path_str()}".encode(
```

Comment:
nit: different sessions will use the same `fragment_id`. But that isn't a problem since its anyways always separated by sessions. So no need to add a session_id here, or?

Reply:
yep -- that should be the case. Each `FragmentStorage` is tied to a session, so any collisions in these hashes shouldn't be an issue assuming the two saved fragments live in different sessions